File: dart/lib/src/utils/derive_address.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:solana/solana.dart';

import '../constants/addresses.dart';
import 'base58_utils.dart';

/// Hash prefix used by the SNS system
const String _hashPrefix = 'SPL Name Service';

/// Generates a SHA-256 hash of a string with the SNS prefix.
///
/// @param str The string to be hashed
/// @returns SHA-256 hash as bytes
Future<Uint8List> generateHash(String str) async {
  final data = utf8.encode(_hashPrefix + str);
  final hash = sha256.convert(data);
  return Uint8List.fromList(hash.bytes);
}

/// Derives an address from a hash with optional parent and class addresses.
///
/// @param hash The hash to derive the address from
/// @param parentAddress Optional parent address (32 bytes, defaults to zero)
/// @param classAddress Optional class address (32 bytes, defaults to zero)
/// @returns The derived address as a base58 string
Future<String> getAddressFromHash(
  Uint8List hash, {
  String? parentAddress,
  String? classAddress,
}) async {
  final seeds = <List<int>>[];

  // Add the hash
  seeds.add(hash);

  // Add class address (or 32 zero bytes)
  if (classAddress != null) {
    seeds.add(_base58Decode(classAddress));
  } else {
    seeds.add(List.filled(32, 0));
  }

  // Add parent address (or 32 zero bytes)
  if (parentAddress != null) {
    seeds.add(_base58Decode(parentAddress));
  } else {
    seeds.add(List.filled(32, 0));
  }

  // Generate PDA with NAME_PROGRAM_ADDRESS
  return _getProgramDerivedAddress(
    seeds: seeds,
    programId: nameProgramAddress,
  );
}

/// Derives an address from an input string with optional parent and class addresses.
///
/// This mirrors the main `deriveAddress` function from js-kit/src/utils/deriveAddress.ts
///
/// [str] - The input string to derive the address from
/// [parentAddress] - Optional parent address
/// [classAddress] - Optional class address
///
/// Returns the derived address as a base58 string
Future<String> deriveAddress(
  String str, {
  String? parentAddress,
  String? classAddress,
}) async {
  final hash = await generateHash(str);
  return getAddressFromHash(
    hash,
    parentAddress: parentAddress,
    classAddress: classAddress,
  );
}

/// Generates a Program Derived Address (PDA) using the official Solana implementation.
///
/// This uses the proven Ed25519HDPublicKey.findProgramAddress from the solana package
/// to ensure 100% compatibility with the official Solana SDK.
///
/// [seeds] - List of seed byte arrays
/// [programId] - The program ID as a base58 string
///
/// Returns the PDA as a base58 string
Future<String> _getProgramDerivedAddress({
  required List<List<int>> seeds,
  required String programId,
}) async {
  // Convert to the format expected by the solana package
  final programKey = Ed25519HDPublicKey.fromBase58(programId);

  // Use the official Solana PDA generation
  final result = await Ed25519HDPublicKey.findProgramAddress(
    seeds: seeds,
    programId: programKey,
  );

  return result.toBase58();
}

/// Decodes a base58 string to bytes using shared utility
List<int> _base58Decode(String input) => Base58Utils.decode(input);

File: js-kit/src/utils/deriveAddress.ts
import {
  Address,
  ReadonlyUint8Array,
  getProgramDerivedAddress,
} from "@solana/kit";

import { addressCodec, utf8Codec } from "../codecs";
import { NAME_PROGRAM_ADDRESS } from "../constants/addresses";

const HASH_PREFIX = "SPL Name Service";

/**
 * Hashes the SPL Name Service hash prefix plus an input string using SHA-256.
 *
 * @param str Input string to hash
 * @returns Hash bytes.
 */
export const _generateHash = async (str: string): Promise<Uint8Array> => {
  const data = utf8Codec.encode(HASH_PREFIX + str);
  const hashBuffer = await crypto.subtle.digest("SHA-256", data);

  return new Uint8Array(hashBuffer);
};

/**
 * Derives an SPL Name Service address from a hash.
 *
 * @param hash Hash bytes used as the first PDA seed
 * @param parentAddress Optional parent address seed
 * @param classAddress Optional class address seed
 * @returns The derived address.
 */
export const _getAddressFromHash = async (
  hash: Uint8Array,
  parentAddress?: Address,
  classAddress?: Address
): Promise<Address> => {
  const seeds: ReadonlyUint8Array[] = [hash];
  seeds.push(
    classAddress ? addressCodec.encode(classAddress) : new Uint8Array(32)
  );
  seeds.push(
    parentAddress ? addressCodec.encode(parentAddress) : new Uint8Array(32)
  );

  const [address] = await getProgramDerivedAddress({
    programAddress: NAME_PROGRAM_ADDRESS,
    seeds,
  });

  return address;
};

/**
 * Derives an address from an input string, with optional parent address and
 * optional class address. These addresses form part of the seeds for Program
 * Derived Address (PDA) derivation.
 *
 * @param str Input string to derive the address from
 * @param parentAddress Optional parent address seed
 * @param classAddress Optional class address seed
 * @returns The derived address.
 */
export const _deriveAddress = async (
  str: string,
  parentAddress?: Address,
  classAddress?: Address
) => {
  const hash = await _generateHash(str);
  const address = await _getAddressFromHash(hash, parentAddress, classAddress);

  return address;
};
